Add functions for flagging auto-correlations

tyler-a-cox opened the pull request:
This PR implements new RFI flagging functions for auto-correlation using channel differencing and DPSS filtering from @AaronParsons [fastcal notebook](https://github.com/HERA-Team/hera_notebook_templates/blob/ef53071cfac322a6fe9d629d0ed508028bd1fe3c/notebooks/fastcal_inspect.ipynb).

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+    data: np.ndarray
+        2D data array of the shape (time, frequency)
+    freqs: np.ndarray
+        1D array of frequencies present in the data
```
Note the proper units of the freqs

tyler-a-cox replied:
Add units to docstring

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+        list of floats of half-widths of delay filter windows in nanosec
+    wgts: np.ndarray
+        2D np.ndarray of real weights. Must be the same shape as data.
+    nsig: float, default=10
```
A default of 10 is different from a lot of the other defaults in this module (usual 5 or 6). Is there a reason for this number?

tyler-a-cox replied:
This is just what Aaron set in the fastcal notebook for the channel difference flagger. I just changed the default value to `nsig=6` and it didn't significantly change the result on the test data. I'll go ahead and set it to 6

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+    sigma = np.median(
+        np.abs(res - np.median(res, keepdims=True, axis=1)), axis=1, keepdims=True,
+    )
+    sigma /= 0.675
```
I'd do 0.6745

tyler-a-cox replied:
Changed this.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+        np.abs(res - np.median(res, keepdims=True, axis=1)), axis=1, keepdims=True,
+    )
+    sigma /= 0.675
+    wgts = np.where(np.abs(res) < nsig * sigma, 1, 0)
```
Why not use the input noise where you have it, instead of 1s?

tyler-a-cox replied:
Changed to use input noise when possible

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+
+    Returns:
+    -------
+    wgts: np.ndarray
```
It's a bit strange that a function called a "flagger" returns floats and not booleans. It's certainly not what I'd expect from the function name...

tyler-a-cox replied:
Changed the name of this to `detrend_dpss_filt` to make it more in line with other flaggers in `xrfi` that do similar things

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+    res = data - model
+
+    # Use noise standard deviation given or estimate it from residuals
+    if noise is not None:
+        sigma = noise
+
+    else:
+        sigma = np.median(
+            np.abs(res - np.median(res, keepdims=True, axis=1)), axis=1, keepdims=True,
+        )
+        sigma /= 0.675
```
If I'm reading this right, the two cases here produce a signal with different shapes. 

Also, what happens if the noise model also has RFI in it?

tyler-a-cox replied:
Removed this

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+        2D data array of the shape (time, frequency)
+    noise: np.ndarray
+        2D array for containing an estimate of the noise standard deviation. Must be the same shape as the data
+    nsig: float, default=10
```
Same comment as above

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+    nsig: float, default=10
+        The number of sigma in the metric above which to flag pixels.
+    kernel_width: list
+        Widths of the convolution kernels used to produce model
```
These are actually (width - 1)/2, which I'd note explicitly in the docstring

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+
+    Returns:
+    -------
+    wgts: np.ndarray
```
Same comment as before about bools

tyler-a-cox replied:
Modified the name of this similarly to above

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+        residual = np.zeros_like(data)
+
+        # For each time integration, identify outliers
+        for t in range(data.shape[0]):
```
Can't this be done in a vectorized way?

tyler-a-cox replied:
I tried modifying this, but didn't find a solution with a much better speed-up. I can try to modify this again if we notice this getting slow.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+        for t in range(data.shape[0]):
+            _d = np.convolve(dw[t], kernel, mode="same") - dw[t]
+            _w = np.convolve(wgts[t], kernel, mode="same") - wgts[t]
+            residual[t] = dw[t] - _d / _w.clip(1, np.inf)
```
What is the clip doing and what do its bounds mean?

tyler-a-cox replied:
The clip is preventing divide by zero errors. The `_w` array shouldn't have values less than one except at the location of flagged channels, which should have zeros.

jsdillon replied:
I still don't get this. Why? Can't one cook up a set of wgts where that wouldn't be true? 

It seems like maybe this function is written assuming that all weights are 1 or zero, but there's no guarantee of that from the docstring or any kind of error handling.

tyler-a-cox replied:
I replaced this with flags instead of weights. I also removed the clipping and went with a robust_divide instead.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+    ----------
+    data: DataContainer
+        DataContainer containing numpy arrays of auto-correlation data with shape (ntimes, nfreqs)
+    flag_method: str
```
I'd suggest picking a default

tyler-a-cox replied:
Picked the `channel_diff_flagger` as a default.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+
+    # Combine into array-wide weights
+    array_weights = np.mean(array_weights, axis=0)
+    array_weights = np.where(array_weights < array_wgts_threshold, 1, 0)
```
If you already know that an antenna is bad, should it count against you here? Like, should the threshold be over all antennas or over the not-already-known-to-be-bad antennas?

jsdillon reviewed (changes requested):
Thanks for this @tyler-a-cox! I have a few questions about the overall design, as well as a handful of minor suggestions/clarifications. Wanna chat later today about this?

tyler-a-cox commented:
Thanks for all of your comments @jsdillon! This should be ready for another look. I also added a function which classifies antennas by their flag fraction relative to the array-wide flag fraction. I added it to `xrfi`, but a better place for it may be `ant_class`. Let me know what you think!

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +225,266 @@ def robust_divide(num, den):
+    array_weights = np.where(array_weights < array_wgts_threshold, 1, 0)
+    return rfi_wgts, array_weights
+
+def auto_rfi_checker(auto_wgts, array_wgts, good=(0, 3), suspect=(3, 6)):
```
To be consistent with the other function, I think we want this to operate on data directly.

jsdillon replied:
Also, I agree that this belongs in ant_class

tyler-a-cox replied:
Moved this to ant_class and made the input `data` like the other functions

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +225,266 @@ def robust_divide(num, den):
+        zero identifying channels found to have RFI
+    array_weights: np.ndarray
+        Contains weights for the entire array averaged over all antennas
+    good: 2-tuple or list of 2-tuple bounds for ranges considered good. In units of relative
```
Outdated docstrings.

tyler-a-cox replied:
Removed this

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -222,6 +225,266 @@ def robust_divide(num, den):
+    if np.isclose(std, 0):
+        std = np.std(flagged_fraction)
+    
+    # For all antennas, determine a z-score
```
One of the ideas behind ant_class is that we're trying to avoid relative cuts and only doing absolute cuts. 

Instead of z-scores, I think we should either cut on A) absolute fraction of the band flagged or B) absolute fraction of the band in excess of the array-averaged flagging. Perhaps you want to play around with a file and figure out which is better for catching outliers and not flagging good stuff?

tyler-a-cox replied:
I ended up changing this to absolute fraction of the band flagged. It seemed to work fairly well at identifying outlier antennas.

jsdillon commented on hera_qm/ant_class.py:
```diff
@@ -320,6 +321,60 @@ def auto_slope_checker(data, good=(-.2, .2), suspect=(-.4, .4), edge_cut=100, fi
+        kernel_widths: list
+            Half-width of the convolution kernels used to produce model. True kernel width is (2 * kernel_width + 1)
+            Only used in the "channel_diff_flagger" step
+        mode: str, default='dpss_solve'
```
Default is out of date

tyler-a-cox replied:
Updated this

jsdillon commented on hera_qm/ant_class.py:
```diff
@@ -320,6 +321,60 @@ def auto_slope_checker(data, good=(-.2, .2), suspect=(-.4, .4), edge_cut=100, fi
     return antenna_bounds_checker(relative_slopes, good=good, suspect=suspect, bad=(-np.inf, np.inf))
 
+def auto_rfi_checker(data, good=(0, 0.1), suspect=(0.1, 0.2), nsig=6, antenna_class=None, flag_threshold=0.1, 
+                     kernel_widths=[3, 4, 5], mode='dpss_matrix', filter_centers=[0, -2700e-9, 2700e-9],
```
These filter centers are a bit odd and really specialized to the H4C system. Do we actually need to include the 2700 ns feature?

tyler-a-cox replied:
Changed this. We can add those options as input as needed

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -425,65 +420,23 @@ def flag_autos(data, flag_method="detrend_diff_filt", nsig=6, int_count=None, an
     # Combine into array-wide weights
     array_weights = np.mean(array_weights, axis=0)
-    array_weights = np.where(array_weights < array_wgts_threshold, 1, 0)
+    array_weights = np.where(array_weights < flag_threshold, False, True)
```
These aren't really weights anymore, no? Probably the docstring should change too.

tyler-a-cox replied:
Changed the names of these

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -328,64 +311,76 @@ def detrend_channel_diff_filt(data, noise, nsig=6, kernel_widths=[1, 2, 4], wgts
-    array_wgts_threshold: float, default=0.2
+    flags: dictionary or DataContainer: default={}
+        Dictionary or DataContainer containing boolean flag arrays. If flags=None, then no data will be masked
+    flag_threshold: float, default=0.1
```
This variable name is a bit misleading, at least to me. I'd call it flag_broadcast_thresh or similar.

tyler-a-cox replied:
Changed the name of this to `flag_broadcast_thresh`

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -276,46 +277,28 @@ def detrend_dpss_filt(data, noise, freqs, filter_centers, filter_half_widths, wg
 
 
-def detrend_channel_diff_filt(data, noise, nsig=6, kernel_widths=[1, 2, 4], wgts=None):
+def channel_diff_flagger(data, noise, nsig=6, kernel_widths=[3, 4, 5], flags=None):
```
You probably want to update the docstring to clarify that it returns flags, not weights

tyler-a-cox replied:
Updated this

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -229,7 +229,7 @@ def robust_divide(num, den):
 # ############################################################################
 
-def detrend_dpss_filt(data, noise, freqs, filter_centers, filter_half_widths, wgts=None,
+def dpss_flagger(data, noise, freqs, filter_centers, filter_half_widths, flags=None,
```
Update the docstring returns to reflect that it returns flags not weights

tyler-a-cox replied:
Updated this

jsdillon reviewed (changes requested):
Nice work! Just a handful of documentation fixes and default questions.

tyler-a-cox commented:
Thanks for the notes @jsdillon. I think this should be the last time you need to look at this. If I missed anything, let me know!
